include/ensmallen_bits/cmaes/cmaes.hpp

```diff
@@ -82,18 +82,18 @@ class CMAES
    * modified to store the finishing point of the algorithm, and the final
    * objective value is returned.
    *
-   * @tparam DecomposableFunctionType Type of the function to be optimized.
+   * @tparam SeparableFunctionType Type of the function to be optimized.
    * @tparam MatType Type of matrix to optimize.
    * @tparam CallbackTypes Types of callback functions.
    * @param function Function to optimize.
    * @param iterate Starting point (will be modified).
    * @param callbacks Callback functions.
    * @return Objective value of the final point.
    */
-  template<typename DecomposableFunctionType,
+  template<typename SeparableFunctionType,
            typename MatType,
            typename... CallbackTypes>
-  typename MatType::elem_type Optimize(DecomposableFunctionType& function,
+  typename MatType::elem_type Optimize(SeparableFunctionType& function,
```

Comment:
Maybe it makes sense to rewrite the calculation of the first objective:

https://github.com/mlpack/ensmallen/blob/9b730c721294db0df7c0b752994b9d255ef58479/include/ensmallen_bits/cmaes/cmaes_impl.hpp#L109-L119

and use the `ArbitraryFunctionTypeAPI` as well. Let me know what you think, I can open a PR.

Reply:
It looks like `RandomSelection` depends on the objective function being separable: https://github.com/mlpack/ensmallen/blob/9b730c721294db0df7c0b752994b9d255ef58479/include/ensmallen_bits/cmaes/random_selection.hpp#L49

So maybe we shouldn't?  I guess another idea would be to make `Function` provide non-separable overloads of `Evaluate()` for separable functions (which I've been meaning to do for a while anyway).

Reply:
You are right, another overload sounds like the right thing to do. I think we should open another PR for that, what do you think?

Reply:
Sure, that works.  Do you want me to make the code changes and open the PR, or do you want to?

Reply:
I can open a PR in the next days, unless you like to, probably not urgent.

Reply:
Sounds good, I'll focus on other PRs like #138.

Reply:
Sounds good.